L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,470 @@
+ public:
+  // Sets the decision of whether to do a crypto check on the server certificates.
+  // The default is true.
+  void set_verify_server_certificates(bool verify_server_certs);
+}
+
+// Builds a ChannelCredentials instance that establishes TLS connections in the manner specified by options.
+std::shared_ptr<ChannelCredentials> TlsCredentials(
+    const TlsChannelCredentialsOptions& options);
+}
+
+// Builds a ServerCredentials instance that establishes TLS connections in the manner specified by options.
+std::shared_ptr<ServerCredentials> TlsServerCredentials(
+    const TlsServerCredentialsOptions& options);
+```
+
+`TlsCredentials` features should represent a superset of `SslCredentials`
+features, so we anticipate `TlsCredentials` being the API of choice now going
+forward.  To make migration from the `SslCredentials` API to the
+`TlsCredentials` API easier, we will also have a method that takes
+`SslCredentialsOptions` and creates a `TlsChannelCredentialsOptions` with the
+same settings. We will do similarly for `TlsServerCredentialsOptions`.
+
+### Credential Reloading
+Credential reloading will be implemented via the `CertificateProviderInterface`.
+Note that there is a naming mismatch here - there exist credentials beyond
+certificates that this _could_ theoretically support.  This provider is
+responsible for sourcing credential material and supplying it to the internal
+stack via the `CertificateProvider`'s `SetRootCertificates` and
+`SetIdentityChainAndPrivateKey` APIs.  Two reference implementations of the
+`CertificateProviderInterface` are provided - `StaticDataCertificateProvider`
+and `FileWatcherCertificateProvider`. The `StaticDataCertificateProvider`
+provides certificates from raw string data. The `FileWatcherCertificateProvider`
+will periodically and asynchronously refresh certificate data from a file,
+allowing changes to certificates without restarting the process. More details
+about these reference implementations are provided below. These implementations
+handle many common use cases. If these providers do not support the intricacies
+for a specific use case, a user can provide their own implementations of the
+`CertificateProviderInterface`.
+
+```c
+/* Opaque types. */
+// A struct that stores the credential data presented to the peer in handshake
+// to show local identity. The private key and certificate chain must be PEM
+// encoded and the public key in the leaf certificate must correspond to the
+// given private key.
+struct IdentityKeyCertPair {
+  std::string private_key_pem;
+  std::string certificate_chain_pem;
+};
+
+/** ------------------------------------ Provider ------------------------------------ */
+/** Provides identity credentials and root certificates.
+*/
+// This represents a potential decoupling of roots and identity chains, with
+// further extension points for something like SPIFFE bundles
+class CertificateProviderInterface {
```

Comment:
Apologies if I'm a broken record with the "watch_root_certs" and "set_root_cert_name" APIs: can we move this functionality to the cert provider rather than having it on the options?

Reply:
I don't believe so - I believe the intention is that a provider could be serving many different sets of creds, thus they need to be separated by user-defined names.
Then the specific `TlsCredentialsOptions` struct used to create a specific `TlsCredentials` needs to know which creds from the provider they need, so the name has to be set there.